pkg/gardenlet/controller/federatedseed/shootstate/shootstate_sync_control.go

```diff
@@ -172,3 +185,37 @@ func (sca *syncControllerArtifacts) shutdownQueues() {
 		artifact.workqueue.ShutDown()
 	}
 }
+
+// ShootRetriever decodes raw Shoots into objects
+type ShootRetriever struct {
+	runtime.Decoder
+}
+
+func newShootRetriever() *ShootRetriever {
+	return &ShootRetriever{serializer.NewCodecFactory(gardenscheme).UniversalDecoder()}
+
+}
+
+// FromCluster retrieves the shoot resource from Cluster resource if any
+func (s *ShootRetriever) FromCluster(ctx context.Context, clusterName string, seedClient kubernetes.Interface) (*gardencorev1beta1.Shoot, error) {
+	cluster := &extensionsv1alpha1.Cluster{}
+
+	if err := seedClient.Client().Get(ctx, kutil.Key(clusterName), cluster); err != nil {
+		return nil, fmt.Errorf("could not get cluster with name %s : %v", clusterName, err)
+	}
+
+	shootInternal := &core.Shoot{}
+	shoot := &gardencorev1beta1.Shoot{}
+
+	if cluster.Spec.Shoot.Raw == nil {
+		return nil, fmt.Errorf("cluster resource %s doesn't contain shoot resource in raw format", clusterName)
+	}
+	if _, _, err := s.Decode(cluster.Spec.Shoot.Raw, nil, shootInternal); err != nil {
```

Comment:
I don't understand - why do you first decode into the internal version and directly after it convert to the `v1beta1`? Can't you directly decode into the `v1beta1`?

Reply:
Copy pasta from the gardener-extenions repo.
I think the idea is if the Shoot is a different version than v1beta1 (e.g. v1beta2) then we won't have to immediately change the code here. But the conversion will only make sense if I extract this method in a utils package and make the gardener-extensions reuse it. Otherwise whoever is promoting the shoot version can update this code as well.

Reply:
For the extensions it's different because they have no control over the version that is used for the shoot. For gardener, the same controller is writing the shoot in a specific version in the `Cluster` resource, hence, it can make assumptions which version to expect and doesn't need this.